boa/src/environment/mod.rs

```diff
@@ -6,3 +6,21 @@ pub mod function_environment_record;
 pub mod global_environment_record;
 pub mod lexical_environment;
 pub mod object_environment_record;
+
+#[derive(Debug)]
+pub enum ErrorKind {
+    ReferenceError(String),
+    TypeError(String),
```

Comment:
I think we could use a `Box<str>` here to reduce a bit the memory footprint.

Reply:
The Error messages are however dynamic (depend on variable names, etc.) which is why we had used a String earlier. Could you please elaborate a bit more?

Reply:
Yep, a `String` contains a pointer, a length and a capacity. The capacity indicates how many bytes are allocated for the string, while the length indicates how many are in use. This is useful to grow the string without needing to allocate, but once we know that the String will not change anymore, it's better to drop the excess capacity and have just a pointer with length, a `Box<str>`.

This can be constructed using [`String::into_boxed_str()`](https://doc.rust-lang.org/stable/std/string/struct.String.html#method.into_boxed_str) you can find more information on the standard library documentation :).

Reply:
Finally, when we are ready to construct the actual Error object, the API currently expects a template which can convert to a String as an argument (for the message).
```Rust
pub fn construct_reference_error<M> (&mut self, msg: M) -> Value
where
    M: Into<String> {
...
}
```
Would a `Box<str>` be useful in this case?

Reply:
Most (if not all) of the things that can be converted to a `String` can be converted to a `Box<str>`, so changing it to `Into<Box<str>>` could make the trick.